### ISSUE TITLE
How to map from m_abilLink to the actual hero ability?

### ISSUE THREAD

**Issue by crorella** [Blizzard/heroprotocol#27]
Hello,

I'm trying to retrieve abilities casted by heroes, in gameevents I see a SCmdEvent is created every time a hero uses an ability:
```
{'_bits': 264,
 '_event': 'NNet.Game.SCmdEvent',
 '_eventid': 27,
 '_gameloop': 108,
 '_userid': {'m_userId': 9},
 'm_abil': {'m_abilCmdData': None, 'm_abilCmdIndex': 0, 'm_abilLink': 173},
 'm_cmdFlags': 256,
 'm_data': {'TargetUnit': {'m_snapshotControlPlayerId': 12,
                           'm_snapshotPoint': {'x': 849167,
                                               'y': 420404,
                                               'z': 32304},
                           'm_snapshotUnitLink': 938,
                           'm_snapshotUpkeepPlayerId': 12,
                           'm_tag': 46137345,
                           'm_targetUnitFlags': 111,
                           'm_timer': 0}},
 'm_otherUnit': None,
 'm_sequence': 1,
 'm_unitGroup': None}
```

Is there any way to determine what ability was the hero using?

Thanks in advance

**Comment by tenarsis** [comment]
That index is a relative index into that specific build's ability data. To map it properly you would either need to read the ability data from that build, or we could add the mapping into the replay file (which could unfortunately increase the replay file size in a non trivial way). Thanks for bringing it up though I don't have much useful suggestions at the moment.

**Comment by crorella** [comment]
Thanks for the reply.
Instead of adding the mappings to each reply, why not just create a file and place it on the website or here at github?
Thanks

**Comment by crorella** [comment]
@tenarsis can you please explain how to get this info?

**Comment by crorella** [comment]
@tenarsis  any update?

**Comment by crorella** [comment]
@tenarsis  I was finally able to get the list of abilities for each build in an automated way. Can you please explain how to create the map?

**Comment by brhs** [comment]
> That index is a relative index into that specific build's ability data. To map it properly you would either need to read the ability data from that build...

Can anyone explain how this is done?

**Comment by Sancus** [comment]
@crorella Did you ever figure out how to relate ability lists to the m_abilLink ID? There's a bunch of parsers that will output ability lists nowadays, but I can't find any information on what the IDs actually signify or how to derive them.

**Comment by healingbrew** [comment]
You basically load up all hero specific data XMLs, you can do this with a CASC library. The files are typically `InternalHeroName + Data.xml` i.e. `AlexstraszaData.xml` 

Navigate to the CHero entry (i.e. `<CHero id="Alexstrasza"/>`), and iterate all HeroAbilArray values. I _believe_  (possibly wrong) that this is what it refers to as m_abilCmdIndex. You can also get the 4 character ID from here (`<AttributeId value="Alex"/>`)

You can get the index (Q/W/E/R1/R2/1/2/3/4/5) from going to the CUnit entries (Hero + hero name, i.e. `<CUnit id="HeroAlexstrsaza">`) and also iterating the `<AlternateUnitArray id="">` values and going to the CUnit entries specified there.  You can get the abilty slots from the `CardLayouts` element.

Additionally, you can get the non-data name by parsing the relative GameStrings.txt and decoding `Button/Name/Button Name` (i.e. `Button/Name/AlexstraszaAbundance`)

**Comment by Sancus** [comment]
Yeah that seems roughly correct, but very roughly. m_AbilLink index numbers do correspond with the order abilities appear in each of the XML files at `mods\heroesdata.stormmod\base.stormdata\gamedata\heroes`.  However, the loading order seems more complex than the list in ConfigData. 

"New" heroes(the ones in `mods/heromods` rather than `mods/heroesdata.stormmod`) seem to be tacked on in a different order. For example, Valeera comes right before Valla('demonhunter'), but her Sinister Strike is `953`. Valla's Q is `307`, Xul, being an old hero who comes after Valla, has the much lower indices, using `490` for his 1(Bone Shield). There are also some generic things not in these XML files, such as Mount(Z) which is 112, and loading order+xml files for these non-specific-hero abilities are not listed in that ConfigData either.

Sure would be nice if we had an easy way to export like SC2 :)

**Comment by MGatner** [comment]
@koliva8245 has done a lot of great work making an easy, automated way of parsing all this with https://github.com/koliva8245/HeroesDataParser. I don't see m_AbilLink specifically but koliva might be willing to explain how to get it from the parsed output and/or add it.

**Comment by koliva8245** [comment]
Sorry can't, no easy way.  I think the way it was done for StarCraft 2 was that the map editor had an export data option or something.  So either blizzard would have to export the data or maybe someone could import all the Hots data in the SC2 editor and export it.
